packages/common/src/utils/get-hub-request-options.ts

```diff
@@ -0,0 +1,53 @@
+/* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
+ * Apache-2.0 */
+
+import { UserSession } from "@esri/arcgis-rest-auth";
+import { getSelf } from "@esri/arcgis-rest-portal";
+import { getHubApiUrlFromPortal } from "../urls/get-hub-api-url-from-portal";
+import { getPortalApiUrl } from "../urls/get-portal-api-url";
+import { IRequestOptions } from "@esri/arcgis-rest-request";
+import { IHubRequestOptionsPortalSelf } from "../types";
+
+export interface IGetHubRequestOptionsOptions extends IRequestOptions {
+  token?: string;
+  portalSelf?: IHubRequestOptionsPortalSelf;
+}
+
+/**
+ * A utility method to create an IHubRequestOptions object
+ * @param {Object} [options={}] An optional map of options
+ * @param {UserSession} [options.authentication] An optional UserSession object
+ * @param {string} [options.portal] An optional portal URL
+ * @param {token} [options.token] An optional token
+ * @param {portalSelf} [options.portalSelf] An optional IHubRequestOptionsPortalSelf object
+ * @returns {Promise}
+ */
+export function getHubRequestOptions(
```

Comment:
From [the authentication docs](https://github.com/Esri/hub.js/blob/master/docs/src/guides/authentication.md),

> Note In the future Hub.js will expose a function to streamline construction of IHubRequestOptions

This method takes a stab at that so we can create an IHubRequestOptions from:
1. `token` & `portal` URL combo
1. existing `UserSession` object
1. `token` & `portalSelf` object combo

since we'll need something like this in `hub-components` since we're likely going to be passing tokens around.

Reply:
@rweber-esri  we should not be working directly with tokens. Our components should take a `session` object when appropriate, and/or a IHubRequestOptions. We will be building an identity component for use outside the Hub application, and that will expose a `session` property as well as raise authentication events.

Reply:
> @rweber-esri we should not be working directly with tokens. Our components should take a `session` object when appropriate, and/or a IHubRequestOptions. We will be building an identity component for use outside the Hub application, and that will expose a `session` property as well as raise authentication events.

@dbouwman
Thanks for the context, I suspected things might change in this department. Would you then say:
1. We do not need this method
2. This method is fine the way it is
3. Refactor this method to accept only a UserSession and optionally an IPortal or IHubRequestOptionsPortalSelf

More of `hub-components` questions, but:
* Do we need to support both `UserSession` and `IHubRequestOptions`? If we pass a `UserSession`, then the component will be responsible for getting an `IHubRequestOptions` for the provided `UserSession`, and we'd likely need to do that in _many_ components. I would argue we should _only_ support passing `IHubRequestOptions` to simplify things unless I'm overlooking (quite possible) a burning need to support both?
* Has there been any discussion surrounding how we plan to support padding `UserSession` or `IHubRequestOptions` from Storybook? We can't pass instances from controls/knobs, out of the box anyways. Maybe a custom decorator or control(s) is in order?

I'm really glad you raised ^, I'm going to update the discussions components _now_ to accept IHubRequestOptions as part of [PR-154](https://github.com/Esri/hub-components/pull/154), as I'm about to add/gate those components in opendata-ui and would prefer to not have to revisit them.

Reply:
@rweber-esri - I think this function is useful, but should just take a UserSession. FWIW - I ended up creating a [similar function](https://github.com/Esri/solution.js/blob/386c9c60f8e72f325a19a1bf7b77c58df9dd3fa5/packages/common/src/create-hub-request-options.ts#L34) in Solution.js and that takes a UserSession, and some solutions-specific stuff b/c there are scenarios with Solutions where we need the portalSelf of a different org.

As for what to pass into components... 100% agree that we don't want every component re-fetching portalSelf and the user. That said I'm not certain how best to proceed - i.e. support session or some "HubSession" object which would basically contain the IHubRequestOptions info, but implemented as a class... this is part of the whole Hub.js "class-like-wrappers" project that never seems to get attention :)

As for passing session into storybook - at this point, that's still an open question. I'm _thinking_ that we need to wrap the components into a `<arcgis-identity>` in each story, but we need to build that component first.

Reply:
Per our side discussion, @dbouwman, I've removed the `getHubRequestOptions` utility method I originally included on this PR since we don't have an immediate need for it, it was out of scope for this story to begin with, and it's implementation likely needed additional consideration.